kvprober: unredact leaseholder information in kvprober logs

VishalJaishankar opened the pull request:
Previously, this is the log line seeing in datadog `...r=393210 having likely leaseholder=‹×› returned success...`

The leaseholder info is not a PII and does not serve any purpose if redacted

This PR makes it redact safe.

Epic: none

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/134617)
<!-- Reviewable:end -->

joshimhoff reviewed (approved):
LGTM! Maybe poke #kv to find a KV reviewer.

kvoli reviewed (commented):
See larger comment, I think someone from @cockroachdb/obs-prs should weigh in. I tagged @dhartunian.

> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/134617#-:-OB_xoL18HWswETnLE6I:buyhv4m)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @VishalJaishankar)

___
*[`-- commits` line 4 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/134617#-OB_wYqf6HPgiTA9_Qpz:-OB_wYqf6HPgiTA9_Qq-:b-z28368):*
nit: the grammar is slightly off, consider changing to "Previously, we would see this log line in redacted output:".
___
*[`-- commits` line 8 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/134617#-OB_wtKW9jon58XEuq8d:-OB_wtKW9jon58XEuq8e:by51ufd):*
nit: Use full sentences, add a period.
___
*[`pkg/kv/kvprober/kvprober.go` line 432 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/134617#-OB_x7m_4jO-XVTwFZxG:-OB_x7m_4jO-XVTwFZxH:bvaitxc) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b73b09eba48326bcf5dae7b8219b7b0d6d441a3/pkg/kv/kvprober/kvprober.go#L432)):*
> ```Go
> 		leaseholder := p.returnLeaseholderInfo(finishAndGetRecording())
> 		ctx = logtags.AddTag(ctx, "leaseholder", leaseholder)
> 		log.Health.Infof(ctx, "kv.Get(%s), r=%v having likely leaseholder=%s returned success in %v", step.Key, step.RangeID, redact.Safe(leaseholder), d)
> ```

I'm not sure marking the arbitrary string returned from a function is future proofing against PII being included. Whilst I think this is safe now, I'd prefer if someone from observability weighs in on the best method of pulling the leaseholder info out of a trace and returning a string (which I think `p.returnLeaseHolderInfo(..)` is doing -- the leaseholder info isn't redacted in the trace, it is due to the conversion to a string iiuc.

@dhartunian would you be able to make a pass? Also see the referenced method:

https://github.com/cockroachdb/cockroach/blob/a8f20b14496db21dce682ca5fb1306c5332cf8cb/pkg/kv/kvprober/kvprober.go#L536-L536


<!-- Sent from Reviewable.io -->

dhartunian commented on pkg/kv/kvprober/kvprober.go:
This is not the correct way to implement safe redaction in this scenario. What you want to do is make `returnLeaseholderInfo` return a redactable object. In this case it's returning a string which previously contained redaction data that was stripped away here: https://github.com/cockroachdb/cockroach/blob/2444856f86e45c16626ac535068b5a1405476325/pkg/kv/kvprober/kvprober.go#L570

This is an unsafe operation and should be modified. What you likely want is for `findLastLogMessage` to return a `RedactableString`. When you pass that string as an arg into `Infof` it will get redacted correctly.

VishalJaishankar replied:
TFTR! I made the change to return a `RedactableString`

VishalJaishankar replied:
@dhartunian I don't understand how this will unredact the nodeId. IIUC, if we don't strip the markers and mark it redact safe, ddog will still show it as redacted, right?

arjunmahishi replied:
@VishalJaishankar You are right. `RedactableString` is an extended string type that is meant to hold a mix of redacted and un-redacted parts. So, just casting the nodeID to this type will not unredact it (because it's always unsafe by default). See [this comment ](https://github.com/cockroachdb/redact/blob/b7e45562c4d0d78dd4d6c7f08f65e8220a0d7b77/internal/markers/markers.go#L24-L26). `RedactableString` are constructed using functions like `redact.Sprint` and `redact.Sprintf`.  This lets users decide which part of the string is safe and which part is unsafe.

Looking at the [doc strings of `redact.Sprintf`](https://github.com/cockroachdb/redact/blob/b7e45562c4d0d78dd4d6c7f08f65e8220a0d7b77/markers_print.go#L33-L39) I would just return `redact.Sprintf(leaseholder[1])`. The 1st arg (`format`) is always considered safe. `redact.Sprint(redact.Safe(leaseholder[1]))` is another way. But not recommended (it's a more expensive operation).

VishalJaishankar replied:
Thanks for the detailed context @arjunmahishi , made the changes, could you PTAL

VishalJaishankar committed:
kvprober: unredact leaseholder information in kvprober logs

Previously, we would see this log line seen in datadog
`...r=393210 having likely leaseholder=‹×› returned success...`

The leaseholder info is not a PII and does not serve any
purpose if redacted.

This PR makes it redact safe.

Epic: none

Release note: None

arjunmahishi reviewed (approved):
LGTM

VishalJaishankar commented on pkg/kv/kvprober/kvprober.go:
```diff
@@ -549,7 +550,7 @@ func (p *Prober) returnLeaseholderInfo(recording tracingpb.Recording) string {
 		// Return leaseholder node ID if found.
 		if len(leaseholder) == 2 {
-			return leaseholder[1]
+			return redact.Sprint(redact.Safe(leaseholder[1]))
```
Thanks! Made the change

VishalJaishankar commented:
bors r=dhartunian

VishalJaishankar commented:
blathers backport 24.3
